# 자바스크립트 12. 프로미스 개념부터 활용까지 JavaScript Promise

Promise

1. Promise is Object for asynchronous operation.
2. state: pending(operation 수행중) -> fulfilled(operation 수행성공) or rejected(operation 실패)
3. Producer: 원하는 기능을 수행하서 원하는 데이터를 만들어냄
4. Consumer: 원하는 데이터를 소비

```jsx
1. Producer
// when new Promise is created, the excutor runs automatically.
const promise = new Promise((resolve, rejcet) => {
	// doing some heavy work (network, read files)
	console.log('doing something....');
	setTimeout(() => {
		resolve('ellie'); // resolve 가 작동함
	}, 2000); // 2초 뒤에 코드가 잘 작동하면
})

// when use 'reject'
const promise = new Promise((resolve, rejcet) => {
	// doing some heavy work (network, read files)
	console.log('doing something....');
	setTimeout(() => {
		reject(new Error('no network')); // Error 라는 객체 사용
		// Error 를 작성할 때는 어떤 에러가 발생했는지 잘 작성해야함
	}, 2000);
})
```

```jsx
2. Consumer: then, catch, finally
// promise 가 잘 작동하면
// .then value의 값을 받아서 기능(함수)으로 전달
promise//
	.then((value) => {
		console.log(value);
	});
// catch 로 에러 다루기
	.catch(error => {
		console.log(error);
	})
// finally, 성공하든 실패하든 실행함
	.finally(() => {
		console.log('finally');
	});
```

```jsx
3. Promise chaining
const fetchNmber = new Promise((resolve, reject) => {
	setTimeout(() => resolve(1), 1000);
});

// .then 에서는 value 값을 바로 전달해도 되고
// 또 다른 비동기인 promise 를 전달해도 됨
fetchNmber // resolve(1)인 1이 전달
.then(num => num * 2); // 1 => 1*2
.then(num => num * 3); // 2 => 2*3
.then(num => { // 6
	return new Promise((resolve, reject) => {
		setTimeout(() => resolve(num - 1), 1000); // 6 - 1
	});
})
.then(num => console.log(num)); // 5 => 5
```

```jsx
4. Error Handling
const getHen = () =>
	new Promise((resolve, reject) => {
		setTimeout(() => resolve('🐔'), 1000);
	});
const getEgg = hen =>
	new Promise((resolve, reject) => {
		setTimeout(() => resolve(`${hen} => 🥚`), 1000);
	});
const cook = egg =>
	new Promise((resolve, reject) => {
		setTimeout(() => resolve(`${egg} => 🍳`), 1000);
	});

// callback 함수를 받아올 때
// 받아온 인자가 1개이고 바로 함수에 1개만 넘길때는
// 인자 생략이 가능함
getHen()
	.then(getEgg)
	.then(cook)
	.then(console.log);
//Promise {<pending>}
//	[[Prototype]]: Promise
//	[[PromiseState]]: "fulfilled"
//	[[PromiseResult]]: undefined
// 🐔 => 🥚 => 🍳

// 중간에 에러가 발생한다면❓❓
const getHen = () =>
	new Promise((resolve, reject) => {
		setTimeout(() => resolve('🐔'), 1000);
	});
const getEgg = hen =>
	new Promise((resolve, reject) => {
		setTimeout(() => reject(new Error(`error! ${hen} => 🥚`)), 1000);
	});
const cook = egg =>
	new Promise((resolve, reject) => {
		setTimeout(() => resolve(`${egg} => 🍳`), 1000);
	});

getHen()
	.then(getEgg)
	.catch(error => {
		return '🍞';
	})
	.then(cook)
	.then(console.log)
	.catch(console.log);

// Promise {<pending>}
// 	[[Prototype]]: Promise
//	[[PromiseState]]: "fulfilled"
//	[[PromiseResult]]: undefined
// 🍞 => 🍳
```

```jsx
5. Callback -> Promise 바꾸기

class UserStorage {
	// 사용자가 로그인을 시도할 때
	loginUser(id, password, onSuccess, onError) {
		setTimeout(() => {
			if(
				(id === 'ellie' && password === 'dream') ||
				(id === 'coder' && password === 'academy')
			) {
				onSuccess(id);
			} else {
				onError(new Error('not found'));
			}
		}, 2000);
	}
	// 사용자가 역활을 받아올 때
	getRoles(user, onSuccess2, onError2) {
		setTimeout(() => {
			if(user === 'ellie') {
				onSuccess2({name: 'ellie', role: 'admin'});
			} else {
				onError2(new Error('no access'));
			}
		}, 1000);
	}
}

const userStorage = new UserStorage();
const id = prompt('enter your id');
const password = prompt('enter your password');
userStorage.loginUser(
	id,
	password,
	user => { // user === onSuccess(id)
		userStorage.getRoles(
			user, // user === id
			userWithRole => { // userWithRole === onSuccess2
				alert(`Hello ${userWithRole.name}, you have a ${userWithRole.role} role`)
			},
			error => {
				console.log(error);
			}
		);
	},
	error => {
		console.log(error);
	}
);

//------------------------------------------------------------
class UserStorage {
	loginUser (id, password) {
		return new Promise((resolve, reject) => {
			setTimeout(() => {
			if(
					(id === 'ellie' && password === 'dream') ||
					(id === 'coder' && password === 'academy')
				) {
					resolve(id); // 로그인에 성공하면 resolve에 id를 인자로 전달
				} else {
					reject(new Error('not found'));
				}
			}, 2000);
		});
	}

	getRoles (user) {
		return new Promise((resolve, reject) => {
			setTimeout(() => {
			if(user === 'ellie') {
				resolve({name: 'ellie', role: 'admin'});
				} else {
					reject(new Error('no access'));
				}
			}, 1000);
		});
	}
}

const userStorage = new UserStorage();
const id = prompt('enter your id');
const password = prompt('enter your password');
userStorage.loginUser(id, password)
	.then(userStorage.getRoles) // user => userStorage.getRoles(user)
	.then(user => alert(`Hello ${user.name}, you have a ${user.role}`))
	.catch(console.log);
```
